### ISSUE TITLE
`torch.nn.utils.clip_grad_norm_()` causes H2D sync with foreach ops.

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#106427]
`clip_grad_norm_()` has an optional `foreach` flag ([here](https://github.com/pytorch/pytorch/blob/ef0576f203bff957d7be976840244f45ff510d2c/torch/nn/utils/clip_grad.py#L14)), that specifies that a foreach kernel should be used.

Unfortunately, that foreach kernel (called [here](https://github.com/pytorch/pytorch/blob/ef0576f203bff957d7be976840244f45ff510d2c/torch/nn/utils/clip_grad.py#L76)) performs a H2D sync:

(1) `clip_coef_clamped` is a scalar-tensor

(2) `aten._foreach_mul_` has a `Scalar` overload and a `TensorList` overload, but not a `Tensor` overload.

(3) `torch._foreach_mul_` coerces the scalar-tensor into a tensor to call the Scalar overload, resulting in an `item()` call (causing a H2D sync).

I had a tentative patch up here to try to force `clip_grad_norm_()` to use the existing `_foreach_mul_.List` overload. It avoids the H2D sync, but ends up dispatching several smaller kernels, which is still not ideal for performance. Instead, we should add a dedicated `_foreach_mul_.Tensor` overload.

**Comment by crcrpar** [comment]
Given the overloads of `_foreach_mul`, IMHO a Tensor as the right-hand side argument can be interpreted as either 0d (scalar) or 1d (scalars like ScalarList).

For now, would you think a Tensor as a scalar would suffice?

**Comment by bdhirsh** [comment]
Hmm, I don't think I follow. Is the suggestion to keep using the `Scalar` overload? I don't think we can do that without forcing a H2D sync.

FWIW, if you pass a 0-dim tensor to an op that has both a scalar and a tensor overload, I think we'll hit the Tensor overload. Tested it out on a random op here:
```
a = torch.ones(4, 4)
b = torch.tensor(2)
torch._add_relu(a,b)  # aten._add_relu has both a Scalar and Tensor overload. This dispatches to the Tensor overload.
```

**Comment by crcrpar** [comment]
oops, apologies for confusion. 

I'm not against adding the overload of Tensor. I'm a bit concerned with the ambiguity of that overload potentially comes with as I think we can pass a Tensor of however many elements. I'm wondering if at the moment you happen to have some preference over how to handle the situation. A scalar Tensor would be quite straightforward. As to another of multiple elements, would we be comfortable with treating it as if it's a ScalarList?

**Comment by bdhirsh** [comment]
Oh I see. Are you referring to a case like this?

```
a, b, c, d = torch.ones(3), torch.ones(4), torch.ones(5), torch.ones(6)
other = torch.ones(4)
torch._foreach_mul([a, b, c, d], other)
```

Where I guess I could imagine two possible outcomes:

(a) we try to compute `[a + other, b + other, c + other, d + other]`. In this case, trying to do this would error: Each elemnt of the first list has a different shape, and `other`'s shape cannot be broadcasted to successfully be added with each element

(b) we try to split `other` apart into four scalars, and do the equivalent of `[a + other[0], b + other[1], c + other[2], d + other[3]]`.

That does seem ambiguous to me. Since for now the only case that has really come up is the "scalar-tensor case", it seems reasonable to me to have the `._foreach_mul_.Tensor` overload expect `other` to be a 0-dim scalar, and raise an error otherwise (so we don't have to immediately decide on the ambiguous behavior).

What do you think?

**Comment by crcrpar** [comment]
Thank you for the concise and detailed explanation. yes, the snippet is exactly what I'm not quite sure about the ground truth behavior. I prefer (b) to (a) since we already have TensorList overload. 
though `_foreach_addcmul.Tensor` is a Tensor representing a ScalarList, which I imagine there could be some confusion in the future (of course we can have detailed and friendly `TORCH_CHECK`)...

**Comment by bdhirsh** [comment]
@crcrpar any interest / bandwidth to take a look at this some time soon? If not no worries - just wanted to check and see if you're interested.

**Comment by crcrpar** [comment]
I guess I will have some time next week

**Comment by janeyx99** [comment]
Ah! I agree there is the ambiguity, but I think that I prefer (a) over (b) because the semantic in b is just using a Tensor as a container and the TensorList overload already covers that case. @crcrpar: you seemed to have preferred (a) over (b) for the same reason though...and I am curious for you to elaborate.

I would bet that the general consensus would pick (a) over (b), and if I am right, I think adding more general Tensor support would be natural here.

**Comment by janeyx99** [comment]
The downside of (a) that you pointed out is reasonable though not necessarily a blocker since we can just have the broadcasting rules kick in.

The other way to do (b) would be to convert each scalar into a Tensor and then pack those tensors in a List, which exists today without any H2D syncs, whereas (a) just doesn't exist today without a H2D sync right? Like in one situation, we're making tensors from scalars, but in the latter, we'd be coercing tensors into scalars?

**Comment by bdhirsh** [comment]
> The downside of (a) that you pointed out is reasonable though not necessarily a blocker since we can just have the broadcasting rules kick in.

Hmm, in the example above:
```
a, b, c, d = torch.ones(3), torch.ones(4), torch.ones(5), torch.ones(6)
other = torch.ones(4)
torch._foreach_mul([a, b, c, d], other)
```

Broadcasting won't help here, since broadcasting rules won't apply for an operation like `torch.mul(a, other)` (multiplying tensors of shapes `(3,)` and `(4,)`). So... (a) is basically useless, unless every tensor in your TensorList has the exact same shape (or a list of shapes that are all broadcastable to a shared shape, like `[(4, 4), (4), (1, 4)]`)

> would be to convert each scalar into a Tensor and then pack those tensors in a List, which exists today without any H2D syncs, whereas (a) just doesn't exist today without a H2D sync right

Yep, this is true (re-using `_foreach_mul.TensorList` today instead of implementing (b)). One counterargument I could imagine: (b) could probably be implemented to be more efficient (and less overhead-bound) than using the `TensorList` overload. And well.. don't foreach ops exist in the first place for perf reasons?

It's hard to say **exactly** how much more efficient without implementing it and doing benchmarks, so maybe that's reason enough to hold off on picking a path? One datapoint we have is that I tried to workaround the `clip_grad_norm()` issue by forcing `clip_grad_norm` to call `_foreach_mul.TensorList` and create a python list of scalar-tensors, and that appeared to be noticeably slow.

**Comment by crcrpar** [comment]
fwiw basically the necessity of broadcast would keep foreach from the fast path unless we're quite certain that the broadcast is an arbitrary shape(s) and one element.

**Comment by janeyx99** [comment]
oh huh how come we cannot parallelize it the same way we parallelize the tensorlist nil tensorlist overload? or is the perf loss from something else?

**Comment by crcrpar** [comment]
https://github.com/pytorch/pytorch/blob/891bb259f86276490e401696d71ec189c003974c/aten/src/ATen/native/ForeachUtils.h#L99-L104
